### DISCUSSION TITLE
[Question] RslRlSymmetryCfg example

### DISCUSSION THREAD

**Discussion by jloganolson** [isaac-sim/IsaacLab#2568] [Q&A]
Has anyone tried using `isaaclab_rl.rsl_rl.RslRlSymmetryCfg` ? 

I couldn't find any examples or documentation beyond this: https://isaac-sim.github.io/IsaacLab/main/source/api/lab_rl/isaaclab_rl.html#isaaclab_rl.rsl_rl.RslRlSymmetryCfg

**Comment by bikcrum** [comment]
@jloganolson 

Yes, I’ve used it. It is fairly straightforward once you understand how the mirroring should apply to your specific robot or actuator configuration.

The key is to implement a `data_augmentation_func` that returns both the original and mirrored versions of the observations and actions. The mirroring logic itself depends on the symmetry of your robot and needs to be implemented manually.

---

### Step 1: Implement the Mirror Functions

```python
def mirror_policy_observation(policy_obs):
    # Your logic to mirror policy observations
    return mirrored_policy_obs

def mirror_critic_observation(critic_obs):
    # Your logic to mirror critic observations
    return mirrored_critic_obs

def mirror_actions(actions):
    # Your logic to mirror actions
    return mirrored_actions
```

---

### Step 2: Define the Data Augmentation Function

```python
def data_augmentation_func(env, obs, actions, obs_type):
    if obs_type == 'policy':
        obs_aug = torch.cat((obs, mirror_policy_observation(obs)), dim=0)
    elif obs_type == 'critic' and obs is not None:
        obs_aug = torch.cat((obs, mirror_critic_observation(obs)), dim=0)
    else:
        raise ValueError(f"Mirror logic for observation type '{obs_type}' not implemented")

    actions_aug = torch.cat((actions, mirror_actions(actions)), dim=0)
    return obs_aug, actions_aug
```

---

### Step 3: Add to PPO Runner Config

```python
@configclass
class MyPPORunnerCfg(RslRlOnPolicyRunnerCfg):
    ...
    algorithm = RslRlPpoAlgorithmCfg(
        ...
        symmetry_cfg = RslRlSymmetryCfg(
            use_data_augmentation=True,
            use_mirror_loss=True,
            mirror_loss_coeff=1.0,
            data_augmentation_func=data_augmentation_func,
        )
    )
```

---

Let me know if you have any specific questions about this.

**Comment by jloganolson** [comment]
This was really helpful, thank you! I stood it up with my G1 23dof in a basic locomotion policy ([source code here](https://github.com/jloganolson/g1_23dof_locomotion_isaac/blob/d6c1c13c03cbe22b1e08bb1fc324e641023c930e/source/g1_23dof_locomotion_isaac/g1_23dof_locomotion_isaac/tasks/manager_based/g1_23dof_locomotion_isaac/agents/rsl_rl_ppo_cfg.py#L120))

One thing I need to dig into further - based on initial results, symmetry loss signficantly improves the policy, but `use_data_augmentation` introduces new asymmetries/artifacts.

**Comment by jloganolson** [comment]
Fixed here https://github.com/jloganolson/g1_23dof_locomotion_isaac/blob/ee072967a7bea4cd6c7c89d64b76a421957245da/source/g1_23dof_locomotion_isaac/g1_23dof_locomotion_isaac/tasks/manager_based/g1_23dof_locomotion_isaac/agents/rsl_rl_ppo_cfg.py#L19
